<a href="https://colab.research.google.com/github/reshminambiar/DW15_Task8/blob/main/Task_8_spelling_correction.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
import numpy as np
import pandas as pd

In [ ]:
pip install textblob
#We take the help of python's textblob package to correct the spelling of words and sentences

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/


##Word Spelling correction

In [ ]:
df = pd.read_table('/content/spelling_correction_words.txt',delimiter=':',usecols=[1], header=None,names=['mis-spelled'])

In [ ]:
df

mis-spelled
0             apeal
1            appple
2          encorage
3         perminant
4     mathematicaly
..              ...
179       nationaly
180           sorce
181         monthes
182     progressess
183             wer

[184 rows x 1 columns]

In [ ]:
from textblob import Word

def correct_word_spelling(word):    
    word = Word(word)    
    result = word.correct()    
    return result

In [ ]:
df['corrected spelling'] = df['mis-spelled'].apply(lambda x:correct_word_spelling(x))

In [ ]:
df

mis-spelled corrected spelling
0             apeal             appeal
1            appple              apple
2          encorage          encourage
3         perminant          perminant
4     mathematicaly       mathematical
..              ...                ...
179       nationaly           national
180           sorce              force
181         monthes             months
182     progressess         progresses
183             wer                her

[184 rows x 2 columns]

##Sentence spelling correction

In [ ]:
from textblob import TextBlob

In [ ]:
with open("sentence.txt", "r") as f:# Opening the text file
    text = f.read()  
    print(text)                   
    textBlb = TextBlob(text)            # Making a textblob
    textCorrected = textBlb.correct()   # Correcting the text
    print(textCorrected)

I really appreciat
I am sory.
How does tat sound?
That sounds grat.
neverr mind.
I leaen English.
I do not understend.
Could you repeet that please?
please talke slower?
Thank you. That helps a lott.
I do not nead a bag.
Can you spel that?
help mi
give a difrent color
I do not kno my size.
Where can I fynd
som one helpp me
Is this delyvered
You have compromised yourself seriously.
I was only Crown Prince then.
I was yung.
I am thirty now
It must be reccovered
We hav tried
I faild.
It must be boght
She will not sel
It is stollen
I will be with you in a momnt.
wen is our meeting
Please cal me
open the leter quickly.
She patted me on the shouldar.
She plays tennis after scool.
Once apon a time
an elderly heron maid his home in a pond ful of fish.
He was stif
he was slow in his old age
he didn't find it easy to catch his lanch.
However, he decyded to use his wits
he went to see a crabb
a great chaterbox
and in a mock frightened volce
told him the lattest rumour.
Certain birds are my frends

In [ ]:
with open("corrected_sentence.txt", "w") as f1:
    f1.write(str(textCorrected))

In [ ]:
#Function to compare two texts and seggregate
def compare(text1, text2):  
    l1 = text1.split()
    l2 = text2.split()
    good = 0
    bad = 0
    for i in range(0, len(l1)):
        if l1[i] != l2[i]:
            bad += 1
        else:
            good += 1
    return (good, bad)

# Helper function to calculate the percentage of misspelled words
def percentageOfBad(x):
    return (x[1] / (x[0] + x[1])) * 100

In [ ]:
with open("sentence.txt", "r") as f1: #typo-filled text
    t1 = f1.read()

with open("sentence_original.txt", "r") as f2: # contains the original text
    t2 = f2.read()

t3 = TextBlob(t1).correct()

mistakesCompOriginal = compare(t1, t2)
originalCompCorrected = compare(t2, t3)

print("Mistakes compared to original ", mistakesCompOriginal)
print("Original compared to corrected ", originalCompCorrected)

print("Percentage of mistakes in the test text: ", percentageOfBad(mistakesCompOriginal), "%")
print("Percentage of mistakes in the corrected text: ", percentageOfBad(originalCompCorrected), "%")

Mistakes compared to original  (371, 99)
Original compared to corrected  (410, 60)
Percentage of mistakes in the test text:  21.06382978723404 %
Percentage of mistakes in the corrected text:  12.76595744680851 %


Here we can see that the mistakes in the text reduced from 21% to 12%. But here there are some limitations as the dataset used here is a manually created and typo filled one. With an original piece of text with spelling mistakes, the error rate will be considerably reduced. 